Adding Software Design Document to mole-docs.readthedocs.io (#352)
csrc-sdsu/mole: MOLE_SW_DESIGN.md

* Begin all source files (including programs, headers, scripts, etc.) with:  
     
  * A comment line describing the contents of the file.

### Directory Structure and Filenames

* File and directory names must contain only lowercase ASCII letters, numbers, underscores, and a period.  The leading character must be alphabetic. Maximum length 31\. Only a single period is permitted.  These requirements ensure file and directory names are relatively portable. No periods for directory names.

**MOLE standard sub-directory names**

| Sub-directory | Contents | Required |
| :---- | :---- | :---- |
| build | CMake files for building library and examples | If any build files. |
| doc | Documentation (HTML,PDF) files. | If several doc files. |
| example | Sample program files, all languages | If several sample files. |
|Julia | Contains Julia's implementation own relevant docs, examples, src, and test subdirectories |
| src | Source files which must be compiled to build the library.  | If any source files. |
| test | Regression or other test programs or scripts. | If several test files. |
 

### Documentation

> Even the simplest library needs some documentation; the amount should be proportional to the need.  The documentation should assume the readers have a basic knowledge of C++, but are not necessarily experts.

Reviewer: Did you revert to an older version? The change Valeria requested is no longer here.

Author: Everything has been incorporated, check the resolved comments above.  Some lines were removed from the document, as they were no longer needed.  Is there a particular section that you don't see.  This is the latest version with all the suggestions incorporated.

Reviewer: Valeria's comment has you changing "C++" to "the implemented language".

Author: I see. . . thanks! I thought I had already changed it but C++ is still there. You mentioned that there were others.  I'll take a look at it later today.  Can you send me other comments/suggestions that need to be added or are missing.  I'll spend sometime on this later today.

* File and directory names must contain only lowercase ASCII letters, numbers, underscores, and a period.  The leading character must be alphabetic. Maximum length 31\. Only a single period is permitted.  These requirements ensure file and directory names are relatively portable. No periods for directory names.

**MOLE standard sub-directory names**

| Sub-directory | Contents | Required |
| :---- | :---- | :---- |
| build | CMake files for building library and examples | If any build files. |
| doc | Documentation (HTML,PDF) files. | If several doc files. |
| example | Sample program files, all languages | If several sample files. |
| src | Source files which must be compiled to build the library.  | If any source files. |
| test | Regression or other test programs or scripts. | If several test files. |

Redirection

If the documentation is in a doc sub-directory, the primary directory index.html file should just do an automatic redirection to the doc subdirectory:

\<html\>  
\<head\>  
\<meta http-equiv="refresh" content="0; URL=doc/index.html"\>  
\</head\>  
\<body\>  
Automatic redirection failed, please go to  
\<a href="doc/index.html"\>doc/index.html\</a\>  
\</body\>  

> \</html\>

Reviewer 2: Is this `html` needed here?

Reviewer 2: Please remove the `html`. Thanks

Reviewer: This is handled by the Sphinx compiler, so it is superfluous here.

Appropriate topics for documentation often include:

* General introduction to the library.  
* Description of each class.  
* Relationship between classes.  
* For each function, as applicable, description, requirements (preconditions), effects, post-conditions, returns, and throws.  
* Discussion of error detection and recovery strategy.  
* How to use it including description of typical uses.  
* How to compile and link.  
* How to test.  
* Version or revision history.  
* Rationale for design decisions.  See [Rationale rationale](#heading=h.bnpcyov6xn2t).  
* Acknowledgements.  See [Acknowledgments rationale.](#heading=h.ehzogyplxsoq)

---

## Rationale

Rationale for some of the requirements and guidelines follows.

### Exception-specification rationale

Exception specifications \[ISO 15.4\] are sometimes coded to indicate what exceptions may be thrown, or because the programmer hopes they will improve performance.  But consider the following member from a smart pointer:

>    T& operator\*() const throw()  { return \*ptr; }

Reviewer 2: Please enclose this in code mode ` `

Author: done!

For example, some compilers turn off inlining if there is an exception-specification.  Some compilers add try/catch blocks. Such pessimizations can be a performance disaster which makes the code unusable in practical applications.

Although initially appealing, an exception-specification tends to have consequences that require **very** careful thought to understand. The biggest problem with exception-specifications is that programmers use them as though they have the effect the programmer would like, instead of the effect they actually have.

A non-inline function is the one place a "throws nothing" exception-specification may have some benefit with some compilers.

---

### Source code fonts rationale 
( from BOOST )

> Dave Abrahams comments: An important purpose (I daresay the primary purpose) of source code is communication: the documentation of intent. This is a doubly important goal for MOLE, I think. Using a fixed-width font allows us to communicate with more people, in more ways (diagrams are possible) right there in the source. Code written for fixed-width fonts using spaces will read reasonably well when viewed with a variable-width font, and as far as I can tell every editor supporting variable-width fonts also supports fixed width. I don't think the converse is true.

Reviewer 2: We should clean up this and avoid direct quotes, I believe

Author: done

Reviewer 2: Hi @Tony-Drummond , thank you, but the paragraph (and the one below) still begins with "Dave Abrahams comments: ". I would personally avoid these. Readers who are not familiar with these authors will be confused.

Reviewer: Same for these, I thought you were removing the quotes, making them more generic, as per Valeria's suggestions.

Author: I removed one or two other quotations but I had left this quote because I thought it is relevant. As it is important to state where some of this ideas have been used (Dave Abrahams + BOOST C++ are relevant in the C++ context), Quotes are used often in rigorous academic arguments. I understand if we want to remove it to make this document more general and less C++ oriented (which font sizes are more general).

## Table of Contents

1. [Introduction](#introduction)
2. [License Requirements](#license-requirements)
3. [Portability requirements](#portability-requirements)
4. [Ownership](#ownership)
5. [Guidelines](#guidelines)
- [Design and Programming](#design-and-programming)
-  [Directory Structure and Filenames](#directory-structure-and-filenames)
-  [Documentation](#documentation)
6. [Rationale](#rationale)
- [Exception-specification rationale](#exception-specification-rationale)
- [Source code fonts rationale](#source-code-fonts-rationale)
- [Tabs rationale](#tabs-rationale)
-  [Rationale rationale](#rationale-rationale)
- [Acknowldegements Rationale](#acknowledgements-rationale )
7. [Naming   consistency](#naming-consistency)

## Introduction

This page describes requirements and guidelines for the content of contributions submitted to MOLE. “Contributions” includes library updates, examples, or anything else under the guise of the code contributed for inclusion in the MOLE Library.  
See the [***Contributing To MOLE*** **webpage**](https://mole-docs.readthedocs.io/en/main/intros/contributing_wrapper.html) for a description of the process involved.  

> Requirements

Reviewer 2: Is this `Requirements` supposed to be a subsection header? Then please use the `###` prefix. Like this, it does not even show on a separate new paragraph. Thanks

To avoid the frustration and wasted time of a proposed contribution being rejected, it must meet these requirements:

* The contribution must be generally useful and not restricted to a narrow problem domain.  

> * The contribution must meet the [portability requirement](#portability-requirements) s below.   

Reviewer 2: ```suggestion
* The contribution must meet the [portability requirements](#portability-requirements) below.   
```

---

## License requirements

MOLE is being developed and distributed under a  GNU GPL v3. See **[LICENSE](https://github.com/csrc-sdsu/mole/blob/main/LICENSE).**

Current requirements for the MOLE  license are: 

* Must be simple to read and understand.  
* Must grant permission without fee to copy, use and modify the software for any use (commercial and non-commercial).   
* Must require that the license appear on all copies of the software source code.  
* Must not require that the license appear with executables or other binary uses of the library.  
* Must not require that the source code be available for execution or other binary uses of the library.  
* Must be compatible with GPL 3.0 license

All Collaborators’ contributions developed and distributed under a different software license not compatible with MOLE’s license will be listed on the [**MOLE website**](http://mole-ose.org) as a recommended reference and include a link where the particular software can be obtained. 

---

## Portability requirements 

> * A contribution interface must be portable and not restricted to a particular compiler or operating system. Ie, A C++ example must not require the Intel Compiler. A Python example does not require a Linux operating system for compiling.  

Reviewer 2: This sentence is a bit tricky. If an user/contributor plans to add a MATLAB/octave contribution, this is only going to work with the MATLAB compiler/interpreter. (similarly, a python/Julia one). But I get the idea of what the opening sentence was trying to convey. It is just the grouping of "compiler or operating system" all in one sentence that makes it a bit not applicable in all cases.

Reviewer 2: Also, I would spell "I.e.," not "Ie". Thank you

> * A contribution's implementation must, if possible, be portable and not restricted to a particular compiler or operating system.  If a portable implementation is not possible, non-portable constructions are acceptable if reasonably easy to port to other environments, and implementations are provided for at least two popular operating systems (such as UNIX and Windows).  

Reviewer 2: This beginning is a repetition (possibly a copy/paste error).

* There is no requirement that a contribution run on C++ compilers that do not conform to the ISO standard. 

Since there is no absolute way to prove portability, many MOLE submissions demonstrate practical portability by compiling and executing correctly with two different C++ compilers, often under different operating systems.  Otherwise reviewers may disbelieve that porting is in fact practical.

---

## Ownership

Are you sure you own the code you are thinking of submitting?   "How to Copyright Software" by MJ Salone, Nolo Press, 1990 says:

> Doing work on your own time that is very similar to programming you do for your employer on company time can raise nasty legal problems.  In this situation, it's best to get a written release from your employer in advance.

Reviewer 2: Not sure this reference is necessary here. I think the purpose of this section is unclear

Author: This came from @jbrzensk.  Jared, if you didn't intend to be in the final document, we can remove it.

Reviewer: I was mashing things together. I do not care if the quote is removed. The line below (line 76) should stay. Some people care about their work and products.

Author: done, thanks!

  * Use fixed-width fonts.  
  * Use spaces rather than tabs.  
  * Limit line lengths to 80 characters.  
* End all documentation files (HTML or otherwise) with a copyright message and a licensing message. See the END\_OF\_THIS\_FILE for an example of the preferred form.  
* Begin all source files (including programs, headers, scripts, etc.) with:  
     
  * A comment line describing the contents of the file.

### Directory Structure and Filenames

* File and directory names must contain only lowercase ASCII letters, numbers, underscores, and a period.  The leading character must be alphabetic. Maximum length 31\. Only a single period is permitted.  These requirements ensure file and directory names are relatively portable. No periods for directory names.

**MOLE standard sub-directory names**

| Sub-directory | Contents | Required |
| :---- | :---- | :---- |
| build | CMake files for building library and examples | If any build files. |
| doc | Documentation (HTML,PDF) files. | If several doc files. |
| example | Sample program files, all languages | If several sample files. |
| src | Source files which must be compiled to build the library.  | If any source files. |

> | test | Regression or other test programs or scripts. | If several test files. |

Reviewer 2: We should mention that now we have a `julia/MOLE.jl` directory with its own relevant `docs`, `examples`, `src`, and `test` subdirectories

Redirection

If the documentation is in a doc sub-directory, the primary directory index.html file should just do an automatic redirection to the doc subdirectory:

```html
\<html\>  
\<head\>  
\<meta http-equiv="refresh" content="0; URL=doc/index.html"\>  
\</head\>  
\<body\>  
Automatic redirection failed, please go to  
\<a href="doc/index.html"\>doc/index.html\</a\>  
\</body\>  
\</html\>
```

### Documentation

> Even the simplest library needs some documentation; the amount should be proportional to the need.  The documentation should assume the readers have a basic knowledge of C++, but are not necessarily experts.

Reviewer 2: ```suggestion
Even the simplest library needs some documentation; the amount should be proportional to the need.  The documentation should assume the readers have a basic knowledge of the implemented language, but are not necessarily experts.
```

> The format for documentation should be HTML, and should not require an advanced browser or server-side extensions. Style sheets are acceptable. ECMAScript/JavaScript is not acceptable. The documentation entry point should always be a file named index.html or index.htm; see [Redirection](#heading=h.f70c0ohjk3o8).

Reviewer 2: ```suggestion
The format for documentation should be Markdown, and should not require an advanced browser or server-side extensions. Style sheets are acceptable. ECMAScript/JavaScript is not acceptable. The documentation entry point should always be a file named index.html or index.htm; see [Redirection](#heading=h.f70c0ohjk3o8).
```

This `Redirection` link does not work. Please remove it.

> There is no single right way to do documentation. HTML documentation is often organized quite differently from traditional printed documents. Task-oriented styles differ from reference oriented styles. In the end, it comes down to the question: Is the documentation sufficient for the mythical "average" C++ programmer to use the library successfully?

Reviewer 2: I would remove this sentence. We have a whole documentation building infrastructure, and we do not write native HTML documentation.

Appropriate topics for documentation often include:

* General introduction to the library.  
* Description of each class.  
* Relationship between classes.  
* For each function, as applicable, description, requirements (preconditions), effects, post-conditions, returns, and throws.  
* Discussion of error detection and recovery strategy.  
* How to use it including description of typical uses.  
* How to compile and link.  
* How to test.  
* Version or revision history.  
* Rationale for design decisions.  See [Rationale rationale](#heading=h.bnpcyov6xn2t).  

> * Acknowledgements.  See [Acknowledgments rationale.](#heading=h.ehzogyplxsoq)

Reviewer 2: These links do not work. Please remove them.

Author: This is an easy fix no need to remove anything

---

## Rationale

Rationale for some of the requirements and guidelines follows.

### Exception-specification rationale

> Exception specifications \[ISO 15.4\] are sometimes coded to indicate what exceptions may be thrown, or because the programmer hopes they will improve performance.  But consider the following member from a smart pointer:

Reviewer 2: ```suggestion
For C++ contributions, exception specifications \[ISO 15.4\] are sometimes coded to indicate what exceptions may be thrown, or because the programmer hopes they will improve performance.  But consider the following member from a smart pointer:
```

( from BOOST )

Dave Abrahams comments: An important purpose (I daresay the primary purpose) of source code is communication: the documentation of intent. This is a doubly important goal for MOLE, I think. Using a fixed-width font allows us to communicate with more people, in more ways (diagrams are possible) right there in the source. Code written for fixed-width fonts using spaces will read reasonably well when viewed with a variable-width font, and as far as I can tell every editor supporting variable-width fonts also supports fixed width. I don't think the converse is true.

---

### Tabs rationale

Tabs are banned because of the practical problems caused by tabs in multi-developer projects, rather than any dislike in principle. Problems include maintenance of a single source file by programmers using tabs and programmers using spaces, and the difficulty of enforcing a consistent tab policy other than just "no tabs". For programming sanity, and compatibility between IDEs, spaces were chosen as the standard.

---

### Rationale rationale 
( from BOOST )

Rationale is defined as "The fundamental reasons for something; basis" by the American Heritage Dictionary.

> Beman Dawes comments:  Failure to supply contemporaneous rationale for design decisions is a major defect in many software projects. Lack of accurate rationale causes issues to be revisited endlessly, causes maintenance bugs when a maintainer changes something without realizing it was done a certain way for some purpose, and shortens the useful lifetime of software.

Reviewer 2: Similar to my comment above. Thank you

Author: Thanks, @valeriabarra. I understand the points you made here. Since @jbrzensk is the author of this file. I will pass the baton to him rather than adding my own interpretation. I thought this had been reviewed and agreed by the team before I came on board. 
I agree that we now need to consider all the new MOLE assets that perhaps were not considered in the original design (e.g., languages = python, julia, fortran, c++ and Octave/MATLAB. resources = we have now a website, also a single script that automatically propagate/updates the license to all files, etc.).  I suggest that we hold on publishing this document until we reach agreement.  Definitively, this is a great start (thank @jbrzensk !)

Reviewer 2: Hi @Tony-Drummond no, it was not reviewed under the same lenses or purpose. Before it was only an internal document. Now we're adding it to our public facing website. So of course the level of scrutiny and review is different :)

Author: I fully agree, and thank you again for starting this process.

Reviewer: I guess I ask if we need outside rationale for our choices? Or, are our choices made, and this is the way? If we are happy with WHY we do things, so we can reword the source code font rationale, remove the rationale rationale, and leave the others as is.

Author: Thank you both, let me try another pass at it, and hopefully captures all of the above.
